### PR TITLE
Add support for generic parameters in call handlers.

### LINKED ISSUE #126: Stack overflow when emulating code which uses generic parameters
When using the new Echo emulator to emulate the following snippet a stack overflow exception is hit when resolving generic parameter types.
```csharp
public interface ITestInterface<out T, in U> {
    public T TestMethod(U u);
}

public class TestClass : ITestInterface<int, long> {
    public static ITestInterface<int, long> Instance = new TestClass();
    
    public int TestMethod(long u) {
        return (int)u;
    }
}

public static class Usage {
    public static int Use() {
        return TestClass.Instance.TestMethod(long.MaxValue);
    }
}
```

Code used to initiate the emulation:
```csharp
var method = module.GetAllTypes().FirstOrDefault(x => x.Name == "Usage")?.Methods.FirstOrDefault(x => x.Name == "Use");

var vm = new CilVirtualMachine(module, false);
var frame = new CallFrame(method, vm.ValueFactory);
vm.CallStack.Push(frame);

vm.Run();
```

Running this code results in one of the following errors occurring:
explorer_iukCU08NNj
explorer_JAWZEeBLou

### PR DESCRIPTION
**Pull request by puff**
Adds handling for generic arguments when calling a method.
~~Fixes~~ Related issue: #126

### CODE REVIEW

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/CallHandlerBase.cs [hunk 1]
```diff
@@ -83,14 +85,25 @@ protected IList<BitVector> GetArguments(CilExecutionContext context, IMethodDesc
...
+                    ? genericContext.GetTypeArgument(genericSentinelParameter)
+                    : sentinelParameter));
+            }
 
             // Pop normal arguments.
             for (int i = method.Signature.ParameterTypes.Count - 1; i >= 0; i--)
-                result.Add(stack.Pop(method.Signature.ParameterTypes[i]));
+            {
+                var parameter = method.Signature.ParameterTypes[i];
+                result.Add(stack.Pop(parameter is GenericParameterSignature genericParameter
```
Comment (ElektroKill) [inline src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/CallHandlerBase.cs:103]: This simple type check will not work for parameter types like `T[]` where `T` is the generic parameter.

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/CallHandlerBase.cs [hunk 2]
```diff
@@ -83,14 +85,25 @@ protected IList<BitVector> GetArguments(CilExecutionContext context, IMethodDesc
...
             var stack = context.CurrentFrame.EvaluationStack;
             var result = new List<BitVector>(method.Signature!.GetTotalParameterCount());
+            var genericContext = GenericContext.FromMethod(method);
             
             // Pop sentinel arguments.
             for (int i = method.Signature.SentinelParameterTypes.Count - 1; i >= 0; i--)
-                result.Add(stack.Pop(method.Signature.SentinelParameterTypes[i]));
+            {
+                var sentinelParameter = method.Signature.SentinelParameterTypes[i];
+                result.Add(stack.Pop(sentinelParameter is GenericParameterSignature genericSentinelParameter
```
Comment (ElektroKill) [inline src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/CallHandlerBase.cs:94]: Same here, the simple type check will not work for parameter types like `T[]` where `T` is the generic parameter.

**Review by ElektroKill — APPROVED** [review]

**Comment by ElektroKill** [comment]
> Adds handling for generic arguments when calling a method. Fixes #126

The PR does not fully fix issue #126, it only fixes a stack overflow issue. The fix uncovered a new issue with the emulator using the exact same test code.

**Comment by Washi1337** [comment]
FYI; Recently two bugs got fixed in AsmResolver that will be present in 5.4.0, fixing a stack overflow related to this. 
- https://github.com/Washi1337/AsmResolver/commit/cc1e79472c46c83162067f9bdccbe2b2d6ce613e
- https://github.com/Washi1337/AsmResolver/commit/a04f4a61ac9a9354506cf2c75ad32709c008f17f

**Pull request merged**